src/Connection/FiniteConnectionPool.php

```diff
@@ -0,0 +1,234 @@
+<?php
+
+namespace Amp\Http\Client\Connection;
+
+use Amp\CancellationToken;
+use Amp\Deferred;
+use Amp\Http\Client\Internal\ForbidSerialization;
+use Amp\Http\Client\InvalidRequestException;
+use Amp\Http\Client\Request;
+use Amp\Http\Client\Response;
+use Amp\Promise;
+use Amp\Success;
+use function Amp\call;
+use function Amp\coroutine;
+
+final class FiniteConnectionPool implements ConnectionPool
+{
+    use ForbidSerialization;
+
+    /** @var int */
+    private $maxConnections;
+
+    /** @var ConnectionFactory */
+    private $connectionFactory;
+
+    /** @var Promise[][] */
+    private $connections = [];
+
+    /** @var Deferred[][] */
+    private $waiting = [];
+
+    /** @var bool[] */
+    private $waitForPriorConnection = [];
+
+    /** @var int */
+    private $totalConnectionAttempts = 0;
+
+    /** @var int */
+    private $totalStreamRequests = 0;
+
+    /** @var int */
+    private $openConnectionCount = 0;
+
+    public function __construct(int $maxConnections, ?ConnectionFactory $connectionFactory = null)
```

Comment:
Is there any relation to per-host connections limit?
Because I think that's the most useful limit - per-authority actually.
Global connection limits aren't much needed I think (but I would happily like to know why I'm wrong, if I am!)

Reply:
The connection-limit is per authority. I could use a static constructor like `LimitedConnectionPool` to make this clearer.